### ISSUE TITLE
Write Room typeclass

### ISSUE THREAD

**Issue by whitenoiseoss** [evennia/ainneve#4]
The base Room typeclass needs to be constructed.

This is a fairly easy task, but it will require knowledge of the Object API. Whoever writes this will be doing things like attributes for a short/long description, a dict or list for contents of the room, perhaps a room size, weight capacity, or something along these lines.

The methods written will do things like dictate how the appearance of a room appears to players (Object's return_appearance method). This is a core piece of functionality, and it must be fleshed out before building can begin.

**Comment by Idioplex** [comment]
I've sketched out a crude representation of the island. I'm no cartographer and the scale is a bit wacky, but I think it should work for development purposes. 
![ainneve_map](https://cloud.githubusercontent.com/assets/5454467/8299138/1fb6d1e8-1942-11e5-8fe1-3c6e3a7a8102.png)

I'll be fleshing out some ideas for the layout of Riverport during the next couple of days,

**Comment by Tirit** [comment]
I have a default room concept I've been working on that I can contribute, and further develop for this issue.

**Comment by time-** [comment]
I'm working on it. George told me to do PR on feature branches and not on master. May you start a branch for room-related features? I wrote a wiki page and I'll prepare basic room struct for tonight I think.
link to wiki: https://github.com/time-/ainneve/blob/rooms/typeclasses/Rooms.md (edit: it's a work in progress)

**Comment by Nadorrano** [comment]
I read the valuable discussion on the IRC log and saw your wiki! nice work! The problem I see, is that some flags are static while other should be dynamic. Also every flag should be aware of the other flags, for compatibility.
I have a quick idea, but I have to think to all consequencies, just a brainstorming. Python can make static things dynamic, can we use this?
Flags in the dict can instead return functions.
For instance IS_DARK is True in your dictionary, but this should be dynamic: a room is not dark if someone have a torch or if it's an external room at 2 p.m. Or can be partially dark if it's an urban sector and the town has lights in the streets.
So:

``` python
flags = {'IS_DARK': is_dark}

def is_dark(self):
  #check stuff and return boolean

dark = flags.get('IS_DARK')()
```

The function can also check "sector" and act consequently
Other flags are static and can simply return a bool without checking nothing

``` python
flags = {'IS_WATER': true}

# helper function, needed because dict always returns a callable
def true():
 return True
```

Maybe the dictionary can be stored somewhere, maybe in world/rules, as a reference for every room.
Each room then simply has a tuple and maybe a method to check flags in dictionary:

``` python
# Room typeclass
self.db.flags = ('IS_DARK', 'IS_WATER')
```

Easy to build online!

Edit:

> Maybe the dictionary can be stored somewhere, maybe in world/rules,

or in the room typeclass, so different room typeclasses have different default flags or just return different functions for the flag
So if I check dict.get('IS_WATER') in the indoor room typeclass, the result is always None/False because such flag should be not present in the class (for example).

**Comment by time-** [comment]
Yes I like it. I try to write down something so you can check it as I don't know if I've understood everything.

**Comment by ergodicbreak** [comment]
@Nadorrano I like where that is going. 

@time- just to clarify, you can develop on your own feature branch, then make a PR against the main Ainneve master.

**Comment by ergodicbreak** [comment]
@time- Griatch made a good comment I just read in IRC log, check out Evennia tags as a possible datastructure for flags/sectors. https://github.com/evennia/evennia/wiki/Tags

**Comment by time-** [comment]
Tags are cool, but we need to hardcode some default tags to be checked by other game mechanics.
We can't allow to tag a room as "obscure" if the look command checks for "is_dark" tag, agree? Any ideas to realize this? I'm thnking about a list and add a @buildtag command that inherit from @tag, but allows only in-list tags (tags could be added later, based on mud-features).
About sectors? Can I go on with sector classes inherited from base room or I move everything on tags? I think using classes is better because we can add some specific attribute to rooms, for example movement points needed to traverse. Sectors will then add some default tags if needed (houses will always have is_indoor tag, while the ocean will always have is_water tag)
While we define these points, I update my wiki, adding any sort of command I think should check on various tags, for example is_dark is checked by look command, while is_water/is_air are checked by at_before_traverse exit's method.

**Comment by Nadorrano** [comment]
Hi, if you still need some advice, I think this class could cover almost any descripted situation, maybe you can try to follow this scheme. If the flag is found the check method return True, but if a property with the same name is found on the parent typeclass its value is returned instead.

``` python
class FlagsHandler(object):
    """
    An handler to manage flags
    """

    def __init__(self, obj):
        #save parent typeclass
        self.obj = obj

    def add(self, flag):
        self.obj.tags.add(flag, category='flags')

    def remove(self, flag):
        if self.obj.tags.get(flag, category='flags'):
            self.obj.tags.remove(flag, category='flags')

    def check(self, flag, default=False):
        if self.obj.tags.get(flag, category='flags'):
            if hasattr(self.obj, flag):
                return getattr(self.obj, flag)
            return True
        return default
```

**Comment by time-** [comment]
@Nadorrano 
awesome! thank you!
I merged your code and fixed some stuff in mine. This handler stuff is very cool.
I want to rewrite the .lists method to avoid copying the @tag code as I did :P but I'll check it later.
For now it works, I fixed the @flag command into build.py, adding handler methods.
As I finish sectors I merge all and PR
https://github.com/time-/ainneve/tree/rooms
I'll fix wiki as last step

**Comment by Nadorrano** [comment]
You can add the handler to the base Object typeclass as a property:

``` python
@property
def flags(self):
    return FlagHandler(self)
```

Then you can access it: `obj.flags.check('is_water')`

pay attention that evennia tags are all lowercase (I think, has to recheck this)

**Comment by time-** [comment]
@Nadorrano all done :) thanks a lot mate.
I also used your tip with sectors. They are not classes anymore but they are set as property. I'm gonna update it with some modifications (default move-points, default capacity, etc) as I was doing with sector classes.
Almost there :D

**Comment by ergodicbreak** [comment]
Closing as this was merged in #22
